File: packages/csharp/RealtimeMail/TrustPolicy.cs
namespace RealtimeMail;

public sealed class TrustPolicy
{
    private readonly HashSet<string> trustedDomains = new(StringComparer.OrdinalIgnoreCase);

    public void TrustDomain(string domain)
    {
        trustedDomains.Add(domain);
    }

    public void RevokeDomain(string domain)
    {
        trustedDomains.Remove(domain);
    }

    public bool IsTrusted(string domain)
    {
        return trustedDomains.Contains(domain);
    }

    public bool CanRender(RealtimeMailMessage message)
    {
        if (message.Source == MailSource.Traditional)
        {
            return message.Capabilities.Contains(TrustCapability.RenderHtml);
        }

        return IsTrusted(message.Domain) && message.Capabilities.Contains(TrustCapability.RenderHtml);
    }

    public bool CanRunScript(RealtimeMailMessage message)
    {
        return message.Source == MailSource.Realtime
            && IsTrusted(message.Domain)
            && message.Capabilities.Contains(TrustCapability.RunScriptSandboxed);
    }
}

File: packages/java/src/main/java/org/realtimemail/TrustPolicy.java
package org.realtimemail;

import java.util.HashSet;
import java.util.Set;

public final class TrustPolicy {
  private final Set<String> trustedDomains = new HashSet<>();

  public void trustDomain(String domain) {
    trustedDomains.add(domain);
  }

  public void revokeDomain(String domain) {
    trustedDomains.remove(domain);
  }

  public boolean isTrusted(String domain) {
    return trustedDomains.contains(domain);
  }

  public boolean canRender(RealtimeMailMessage message) {
    if (message.source() == MailSource.TRADITIONAL) {
      return message.capabilities().contains(TrustCapability.RENDER_HTML);
    }
    return isTrusted(message.domain()) && message.capabilities().contains(TrustCapability.RENDER_HTML);
  }

  public boolean canRunScript(RealtimeMailMessage message) {
    return message.source() == MailSource.REALTIME
      && isTrusted(message.domain())
      && message.capabilities().contains(TrustCapability.RUN_SCRIPT_SANDBOXED);
  }
}
